File: 1-Two-Sum.erl
-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
    L=lists:seq(0,length(Nums)-1),
    N=lists:zip(L,Nums),
    Ti = [[Xi,Yi] || {Xi,X}<-N,{Yi,Y}<- N--[{Xi,X}],X+Y == Target],
    hd(Ti).

File: 1-Two-Sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    find_two_sum(nums, target)
  end

  def find_two_sum([head|tail], target, position \\\\ 0, seen \\\\ %{}) do
    if Map.has_key?(seen, target-head) do
      [Map.get(seen, target-head), position]
    else
      find_two_sum(tail, target, position+1, Map.put_new(seen, head, position))
    end
  end
end